bip-ccv.mediawiki

```diff
@@ -0,0 +1,420 @@
+
+=== Script support for <code>OP_CHECKCONTRACTVERIFY</code> ===
+
+The specification is divided in three parts:
+* the transaction-wide initialization;
+* the input initialization;
+* the opcode evaluation.
+
+The following helper function is a version of <code>taproot_tweak_pubkey</code>, except that a raw 32-byte data is used as the tweak.
+
+<source lang="python">
+def tweak_embed_data(pubkey, data):
+    assert len(pubkey) == 32
+
+    data_tweak = sha256(pubkey + data)
+
+    t = int_from_bytes(data_tweak)
+    if t >= SECP256K1_ORDER:
+        raise ValueError
+    P = lift_x(int_from_bytes(pubkey))
+    if P is None:
+        raise ValueError
+    Q = point_add(P, point_mul(G, t))
+    return 0 if has_even_y(Q) else 1, bytes_from_int(x(Q))
+</source>
+
+The <code>taproot_tweak_pubkey</code> from [[bip-0341.mediawiki|BIP341]] is also used as a helper function.
+
+The following notations are used in the pseudocode below:
+* <code>n_inputs</code> and <code>n_outputs</code> are the number of inputs and outputs of the transaction, respectively;
+* <code>inputs[i]</code> is the i-th input of the transaction;
+* <code>outputs[i]</code> is the i-th output of the transaction;
+* <code>this_input_index</code>, <code>this_input_internal_key</code> and <code>this_input_taptree</code> are the index, taproot internal key and taproot Merkle root of the current input, respectively.
+* <code>P2TR(key)</code> computes the scriptPubKey of the P2TR output with the given key as the taproot output key.
+
+==== Transaction-wide initialization ====
+
+This is executed once for the entire transaction, before any of the transaction input's Scripts are evaluated. It initializes three arrays that are used to keep track of the amount information of the output.
+
+<source lang="python">
+  output_min_amount = [0] * n_outputs
+  output_checked_default = [False] * n_outputs
+  output_checked_deduct = [False] * n_outputs
+</source>
+
+==== Input initialization ====
+
+This is executed at the beginning of the evaluation of each input's Script. It initializes the residual amount to equal the full amount of the current input.
+
+<source lang="python">
+  residual_input_amount = input[this_input_index].amount
+</source>
+
+==== <code>OP_CHECKCONTRACTVERIFY</code> evaluation ====
+
+The following code is executed every time the <code>OP_CHECKCONTRACTVERIFY</code> opcode is encountered during the evalutation of a taproot Script spend. <code>this_input_index</code>, <code>this_input_internal_key</code> and <code>this_input_taptree</code> are the index, taproot internal key and taproot Merkle root of the current input.
+
+<code>BIP341_NUMS_KEY</code> is the x-only provably unspendable key <code>50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0</code> defined in [[bip-341.mediawiki|BIP-341]].
+
+<source lang="python">
+  if flags < CCV_FLAG_CHECK_INPUT or flags > CCV_FLAG_DEDUCT_OUTPUT_AMOUNT:
```

Comment:
I assume you mean to check that no unknown flags are in use? This seems wrong though, because `CCV_FLAG_DEDUCT_OUTPUT_AMOUNT | CCV_FLAG_IGNORE_OUTPUT_AMOUNT > CCV_FLAG_DEDUCT_OUTPUT_AMOUNT`

Probably more clear to have something like `ALL_FLAGS = ... | CCV_FLAG_DEDUCT_OUTPUT_AMOUNT | CCV_FLAG_IGNORE_OUTPUT_AMOUNT | ...` and check `flags && !ALL_FLAGS == 0`

Reply:
Since flags are mutually exclusive and the defined values are -1, 0, 1, 2, this is just checking if you're out of the range (then, _SUCCESS_ as an upgrade hook to add future flags).

Reply:
See above comment, I think calling it "mode" would make this code clear:

```
if mode < CCV_FLAG_CHECK_INPUT or mode > CCV_FLAG_DEDUCT_OUTPUT_AMOUNT:
```

Reply:
Changed to `mode` instead of `flags` in 4ed71843e51742b6a60f445c272b73a1e90dc10f.